docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,706 @@
+
+In function `applyDropTable`, we will delete the table's foreign key information from `infoSchema.referredForeignKeyMap`.
+
+In function `applyCreateTable`, we will add the table's foreign key information into `infoSchema.referredForeignKeyMap` first, 
+then get the table's `ReferredFKInfo` by schema name and table name, then store the `ReferredFKInfo` into `TableInfo.ReferredForeignKeys`.
+
+Then `applyTableUpdate` will also need to reload the old/new table's referred table information, also uses `applyDropTable` to drop the old reference table, use `applyCreateTable` to create new reference table.
+
+That's all.
+
+### Alter Table Add Foreign Key
+
+Here is an example:
+
+```sql
+create table t1 (id int key,a int, index(a));
+create table t2 (id int key,a int);
+alter  table t2 add foreign key fk(a) references t1(id) ON DELETE CASCADE;
+```
+
+Just like create table, we should validate first, and return error if the conditions for creating foreign keys are not met, and also need to do double-check.
+
+When build `TableInfo`, we need to auto create an index for foreign key columns if there is no index cover foreign key columns.
+And this is divides the problem into two cases:
+- Case-1: No need to create index automatically, and only add foreign key constraint.
+- Case-2: Need auto create index for foreign key
+
+#### Case-1: Only add foreign key constrain
+
+The DDL owner handle add foreign key constrain step is:
+
+1. None -> Write Only: add foreign key constrain which state is `write-only` into table.
+3. Write Only - Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:
+   ```sql
+   select 1 from t2 where t2.a is not null and t2.a not in (select id from t1) limit 1;
+   ```
+   The expected result is `empty`, otherwise, an error is returned and cancel the ddl job.
+4. Write Reorg -> Public: update the foreign key constrain state to `public`.
+
+A problem is, How the DML treat the foreign key with on delete/update cascade behaviour which state is non-public?
+Here is an example:
+
+```sql
+create table t1 (id int key,a int, index(a));
+create table t2 (id int key,a int, index(a));
+insert into t1 values (1,1);
+insert into t2 values (1,1);
+alter  table t2 add constraint fk_1 foreign key (a) references t1(id) ON DELETE CASCADE;
+```
+
+The schema change of foreign key `fk_1` is from `None` -> `Write-Only` -> `Write-Reorg` -> `Public`。
+When the foreign key `fk_1` in `Write-Only` state, a DML request has come to be processed:
+
+```sql
+delete from t1 where id = 1;
+```
+
+Then, TiDB shouldn't do cascade delete for foreign key `fk_1` in state `Write-Only`, since the `Add Foreign Key` DDL job maybe
+failed in `Write-Reorg` state and rollback the DDL job. But it is hard to rollback the cascade deleted executed before.
+
+So, when execute DML with `non-publick` foreign key, TiDB will do foreign key constraint check instead of foreign key cascade behaviour.
```

Comment:
```suggestion
So, when execute DML with `non-public` foreign key, TiDB will do foreign key constraint check instead of foreign key cascade behaviour.
```